File: carui-frontend/src/services/speed/geo.ts
const EARTH_RADIUS_M = 6371000.0;
const EARTH_RADIUS_KM = 6371.0;

export function haversineM(
	lat1: number,
	lon1: number,
	lat2: number,
	lon2: number,
): number {
	const dlat = ((lat2 - lat1) * Math.PI) / 180;
	const dlon = ((lon2 - lon1) * Math.PI) / 180;
	const a =
		Math.sin(dlat / 2) ** 2 +
		Math.cos((lat1 * Math.PI) / 180) *
			Math.cos((lat2 * Math.PI) / 180) *
			Math.sin(dlon / 2) ** 2;
	return EARTH_RADIUS_M * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

export function haversineKm(
	lat1: number,
	lon1: number,
	lat2: number,
	lon2: number,
): number {
	return haversineM(lat1, lon1, lat2, lon2) / 1000;
}

export function bearing(
	lat1: number,
	lon1: number,
	lat2: number,
	lon2: number,
): number {
	const lat1r = (lat1 * Math.PI) / 180;
	const lat2r = (lat2 * Math.PI) / 180;
	const dlon = ((lon2 - lon1) * Math.PI) / 180;
	const x = Math.sin(dlon) * Math.cos(lat2r);
	const y =
		Math.cos(lat1r) * Math.sin(lat2r) -
		Math.sin(lat1r) * Math.cos(lat2r) * Math.cos(dlon);
	return ((Math.atan2(x, y) * 180) / Math.PI + 360) % 360;
}

export function destinationPoint(
	lat: number,
	lon: number,
	bearingDeg: number,
	distKm: number,
): [number, number] {
	const d = distKm / EARTH_RADIUS_KM;
	const brng = (bearingDeg * Math.PI) / 180;
	const lat1 = (lat * Math.PI) / 180;
	const lon1 = (lon * Math.PI) / 180;
	const lat2 = Math.asin(
		Math.sin(lat1) * Math.cos(d) +
			Math.cos(lat1) * Math.sin(d) * Math.cos(brng),
	);
	const lon2 =
		lon1 +
		Math.atan2(
			Math.sin(brng) * Math.sin(d) * Math.cos(lat1),
			Math.cos(d) - Math.sin(lat1) * Math.sin(lat2),
		);
	return [(lat2 * 180) / Math.PI, (lon2 * 180) / Math.PI];
}

export function angleDiff(a: number, b: number): number {
	const d = Math.abs(a - b);
	return d > 180 ? 360 - d : d;
}

export function pointToSegmentDistance(
	lat: number,
	lon: number,
	segLat1: number,
	segLon1: number,
	segLat2: number,
	segLon2: number,
): number {
	const ax = lat - segLat1;
	const ay = lon - segLon1;
	const cx = segLat2 - segLat1;
	const cy = segLon2 - segLon1;
	const dot = ax * cx + ay * cy;
	const lenSq = cx * cx + cy * cy;
	const t = lenSq !== 0 ? Math.min(1, Math.max(0, dot / lenSq)) : 0;
	const projLat = segLat1 + t * cx;
	const projLon = segLon1 + t * cy;
	return haversineM(lat, lon, projLat, projLon);
}

export function pointToPolylineDistance(
	lat: number,
	lon: number,
	polyline: Array<[number, number]>,
): number {
	let min = Number.MAX_VALUE;
	for (let i = 0; i < polyline.length - 1; i++) {
		const d = pointToSegmentDistance(
			lat,
			lon,
			polyline[i][0],
			polyline[i][1],
			polyline[i + 1][0],
			polyline[i + 1][1],
		);
		if (d < min) min = d;
	}
	return min;
}

export function polylineAzimuth(polyline: Array<[number, number]>): number {
	if (polyline.length < 2) return 0;
	const first = polyline[0];
	const last = polyline[polyline.length - 1];
	return bearing(first[0], first[1], last[0], last[1]);
}

export function latLonToTile(
	lat: number,
	lon: number,
	level: number,
): [number, number] {
	const size = 180 / (1 << level);
	const x = Math.floor((lon + 180) / size);
	const y = Math.floor((lat + 90) / size);
	return [x, y];
}

export function tileSizeDeg(level: number): number {
	return 180 / (1 << level);
}

File: carui-backend/crates/carui-common/src/geo.rs


use std::f64::consts::PI;

const EARTH_RADIUS_M: f64 = 6_371_000.0;
const EARTH_RADIUS_KM: f64 = 6_371.0;


pub fn haversine_m(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    let dlat = (lat2 - lat1).to_radians();
    let dlon = (lon2 - lon1).to_radians();

    let a = (dlat / 2.0).sin().powi(2)
        + lat1.to_radians().cos() * lat2.to_radians().cos() * (dlon / 2.0).sin().powi(2);

    EARTH_RADIUS_M * 2.0 * a.sqrt().atan2((1.0 - a).sqrt())
}


pub fn haversine_km(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    haversine_m(lat1, lon1, lat2, lon2) / 1000.0
}


pub fn bearing(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    let lat1 = lat1.to_radians();
    let lat2 = lat2.to_radians();
    let dlon = (lon2 - lon1).to_radians();

    let x = dlon.sin() * lat2.cos();
    let y = lat1.cos() * lat2.sin() - lat1.sin() * lat2.cos() * dlon.cos();

    (x.atan2(y).to_degrees() + 360.0) % 360.0
}


pub fn destination_point(lat: f64, lon: f64, bearing_deg: f32, dist_km: f64) -> (f64, f64) {
    let d = dist_km / EARTH_RADIUS_KM;
    let brng = (bearing_deg as f64).to_radians();
    let lat1 = lat.to_radians();
    let lon1 = lon.to_radians();

    let lat2 = (lat1.sin() * d.cos() + lat1.cos() * d.sin() * brng.cos()).asin();
    let lon2 = lon1 + (brng.sin() * d.sin() * lat1.cos()).atan2(d.cos() - lat1.sin() * lat2.sin());

    (lat2.to_degrees(), lon2.to_degrees())
}


pub fn angle_diff(a: f64, b: f64) -> f64 {
    let d = (a - b).abs();
    if d > 180.0 {
        360.0 - d
    } else {
        d
    }
}


pub fn point_to_segment_distance(
    lat: f64,
    lon: f64,
    seg_lat1: f64,
    seg_lon1: f64,
    seg_lat2: f64,
    seg_lon2: f64,
) -> f64 {
    let (a, b) = (lat - seg_lat1, lon - seg_lon1);
    let (c, d) = (seg_lat2 - seg_lat1, seg_lon2 - seg_lon1);

    let dot = a * c + b * d;
    let len_sq = c * c + d * d;

    let t = if len_sq != 0.0 {
        (dot / len_sq).clamp(0.0, 1.0)
    } else {
        0.0
    };

    let proj_lat = seg_lat1 + t * c;
    let proj_lon = seg_lon1 + t * d;

    haversine_m(lat, lon, proj_lat, proj_lon)
}


pub fn point_to_polyline_distance(lat: f64, lon: f64, polyline: &[(f64, f64)]) -> f64 {
    polyline
        .windows(2)
        .map(|w| point_to_segment_distance(lat, lon, w[0].0, w[0].1, w[1].0, w[1].1))
        .fold(f64::MAX, f64::min)
}


pub fn polyline_azimuth(polyline: &[(f64, f64)]) -> f64 {
    if polyline.len() < 2 {
        return 0.0;
    }
    let first = polyline.first().unwrap();
    let last = polyline.last().unwrap();
    bearing(first.0, first.1, last.0, last.1)
}


pub fn lat_lon_to_tile(lat: f64, lon: f64, level: i32) -> (i32, i32) {
    let size = 180.0 / (1 << level) as f64;
    let x = ((lon + 180.0) / size) as i32;
    let y = ((lat + 90.0) / size) as i32;
    (x, y)
}


pub fn tile_size_deg(level: i32) -> f64 {
    180.0 / (1 << level) as f64
}


pub fn bounding_box(points: &[(f64, f64)]) -> Option<(f64, f64, f64, f64)> {
    if points.is_empty() {
        return None;
    }

    let mut min_lat = f64::MAX;
    let mut max_lat = f64::MIN;
    let mut min_lon = f64::MAX;
    let mut max_lon = f64::MIN;

    for &(lat, lon) in points {
        min_lat = min_lat.min(lat);
        max_lat = max_lat.max(lat);
        min_lon = min_lon.min(lon);
        max_lon = max_lon.max(lon);
    }

    Some((min_lat, max_lat, min_lon, max_lon))
}


#[inline]
pub fn deg_to_rad(deg: f64) -> f64 {
    deg * PI / 180.0
}


#[inline]
pub fn rad_to_deg(rad: f64) -> f64 {
    rad * 180.0 / PI
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_haversine() {
        
        let dist = haversine_km(55.7558, 37.6173, 59.9343, 30.3351);
        assert!((dist - 634.0).abs() < 10.0);
    }

    #[test]
    fn test_bearing() {
        
        let b = bearing(0.0, 0.0, 1.0, 0.0);
        assert!((b - 0.0).abs() < 1.0);

        
        let b = bearing(0.0, 0.0, 0.0, 1.0);
        assert!((b - 90.0).abs() < 1.0);
    }

    #[test]
    fn test_angle_diff() {
        assert!((angle_diff(10.0, 350.0) - 20.0).abs() < 0.001);
        assert!((angle_diff(0.0, 180.0) - 180.0).abs() < 0.001);
    }
}
